Frameworks and Carthage Support

sagmor opened the pull request:
I just added Framework targets for both Mac and iOS.

I named the Framework GCDWebServers (with an s) because Included frameworks can only `#include <>` the umbrella header, which means that you can't pick individual headers and since EVERY `#include` has to be on the umbrella header it would have bloated `GCDWebServer.h`

Instead with the GCDWebServers name, it's implicit, that the framework includes all the Servers and we keep a clean `GCDWebServer.h`

Now, having framework shared targets in the project, means that this library should support Carthage out of the box (which solves #136) yet for full support, it might be a good idea to provide a prebuilt framework on version releases. It's quite simple so I'll take a look at how to automate that and I'll get back on that later.

I'm currently working with the Mac version on an App I'm working on, but it might need a little more testing to be sure everything works as it should.

I'm leaving this as a PR so we don't duplicate efforts with #161 and to be sure you like this approach before investing more time on it :beers:

swisspol commented:
Thanks for doing this. For iOS you mean a static library right? Not a framework.

> On Apr 16, 2015, at 12:24 PM, Seba Gamboa <notifications@github.com> wrote:
> 
> I just added Framework targets for both Mac and iOS.
> 
> I named the Framework GCDWebServers (with an s) because Included frameworks can only #include <> the umbrella header, which means that you can't pick individual headers and since EVERY #include has to be on the umbrella header it would have bloated GCDWebServer.h
> 
> Instead with the GCDWebServers name, it's implicit, that the framework includes all the Servers and we keep a clean GCDWebServer.h
> 
> Now, having framework shared targets in the project, means that this library should support Carthage out of the box (which solves #136) yet for full support, it might be a good idea to provide a prebuilt framework on version releases. It's quite simple so I'll take a look at how to automate that and I'll get back on that later.
> 
> I'm currently working with the Mac version on an App I'm working on, but it might need a little more testing to be sure everything works as it should.
> 
> I'm leaving this as a PR so we don't duplicate efforts with #161 and to be sure you like this approach before investing more time on it 
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/swisspol/GCDWebServer/pull/166
> 
> Commit Summary
> 
> Generating Frameworks
> Fixed buildable names
> Forgot some '>'s
> File Changes
> 
> M GCDWebServer.xcodeproj/project.pbxproj (514)
> A GCDWebServer.xcodeproj/xcshareddata/xcschemes/GCDWebServers (Mac).xcscheme (110)
> A GCDWebServer.xcodeproj/xcshareddata/xcschemes/GCDWebServers (iOS).xcscheme (77)
> A GCDWebServer/Framework/GCDWebServers.h (56)
> A GCDWebServer/Framework/Info.plist (26)
> Patch Links:
> 
> https://github.com/swisspol/GCDWebServer/pull/166.patch
> https://github.com/swisspol/GCDWebServer/pull/166.diff
> —
> Reply to this email directly or view it on GitHub.
>

sagmor commented:
It's an iOS Framework.

They are supported since iOS 8

swisspol commented on GCDWebServer/Framework/Info.plist:
```diff
@@ -0,0 +1,26 @@
+	<string>FMWK</string>
+	<key>CFBundleShortVersionString</key>
+	<string>3.2.3</string>
+	<key>CFBundleSignature</key>
```
`CFBundleSignature` can be removed

swisspol commented on GCDWebServer/Framework/Info.plist:
```diff
@@ -0,0 +1,26 @@
+	<string>????</string>
+	<key>CFBundleVersion</key>
+	<string>$(CURRENT_PROJECT_VERSION)</string>
+	<key>NSPrincipalClass</key>
```
`NSPrincipalClass` can be removed too

swisspol commented on GCDWebServer/Framework/Info.plist:
```diff
@@ -0,0 +1,26 @@
+	<key>CFBundlePackageType</key>
+	<string>FMWK</string>
+	<key>CFBundleShortVersionString</key>
+	<string>3.2.3</string>
```
Since `CFBundleVersion` is coming from build settings, `CFBundleShortVersionString` should come as well.

swisspol commented:
> Instead with the GCDWebServers name, it's implicit, that the framework includes all the Servers and we keep a clean GCDWebServer.h

That's a good idea!

> I'm leaving this as a PR so we don't duplicate efforts with #161 and to be sure you like this approach before investing more time on it

I just closed #161 as this approach sounds better.

sagmor commented:
Great, I'll remove those keys and setup a way of keeping the version number on a single spot (shared with the Cocoa Pod version)

sagmor commented:
I removed all the unnecessary keys and fixed a minor iOS build error.

The only thing that I'm not too sure is about CFBundleShortVersionString.
Looking at other frameworks. Most of them leave it as 1.0 ([ReactiveCocoa](https://github.com/ReactiveCocoa/ReactiveCocoa/blob/fa50eb340107bea8ac7fde189d9e47594c07924d/ReactiveCocoa.xcodeproj/xcshareddata/xcschemes/ReactiveCocoa%20iOS.xcscheme) for example) and just handle versions with Git's tags ([ReactiveCocoa](https://github.com/ReactiveCocoa/ReactiveCocoa/releases) again).

It seams to be the best approach, but I could take a further look at it :smile:

swisspol commented:
I suspected as such: let's delete the `CFBundleShortVersionString` entirely then.

sagmor commented:
I removed it,
I thought it was required, but it seams to be working without it (at least in development).

sagmor commented:
There it is, I fixed the `APPLICATION_EXTENSION_API_ONLY` error and added both frameworks as dependencies so they get compiled by travis to check for errors :+1:

sagmor commented:
I just added a release script.
```bash
$ ./Archive.sh
```

It uses carthage to build the framework into `GCDWebServers.framework.zip`
That file can be attached to [GCDWebServer's releases](https://github.com/swisspol/GCDWebServer/releases) and comes in the format carthage expects to import dependencies or can be downloaded to be manually imported into projects (check [objective-git](https://github.com/libgit2/objective-git/releases) for example).

swisspol commented on GCDWebUploader/GCDWebUploader.m:
```diff
@@ -296,7 +296,13 @@ @implementation GCDWebUploader
-    NSBundle* siteBundle = [NSBundle bundleWithPath:[[NSBundle mainBundle] pathForResource:@"GCDWebUploader" ofType:@"bundle"]];
+    NSString *bundlePath = [[NSBundle mainBundle] pathForResource:@"GCDWebUploader"
+                                                           ofType:@"bundle"];
+    if (!bundlePath) {
```
You don't need this as `[[NSBundle bundleForClass:[GCDWebUploader class]] ` should work in both cases.

sagmor replied:
I kept it because it leave the door open for changing `GCDWebUploader.bundle` so it search first on the main bundle (so a developer can change the bundle) and then on the framework's resources for the default one

swisspol replied:
Sorry but I don't understand what you mean here.

sagmor replied:
Let's say you want to use a custom GCDWebUploader.bundle (change the CSS to match your app design for example) if I force `[[NSBundle bundleForClass:[GCDWebUploader class]]` you won't be able to change it without touching the .framework directory.

With the current design it first searches for the bundle on the application bundle and then on the framework's bundle.

Any way, it could be removed, and maybe later add a more explicit way of setting the `GCDWebUploader.bundle` in case you want to customize it.

swisspol replied:
OK I understand now. Nobody has ever asked for this however, and even then it's questionable that's the right approach. Think about it, now you have 2 copies of that bundle in the app, and it's very easy for them to get out of sync. The bundle is expected to be in sync with the code that uses it.

If you want to customize the CSS, it's much safer to fork the project. Please remove this commit.

swisspol commented on GCDWebServer/Framework/GCDWebServers.h:
```diff
@@ -0,0 +1,56 @@
+#import <Foundation/Foundation.h>
+
+//! Project version number for GCDWebServers.
+FOUNDATION_EXPORT double GCDWebServersVersionNumber;
```
Where is this symbol coming from?

sagmor replied:
They are generated internally by XCode at compile time.

swisspol replied:
You should remove these 2 symbols from the header. Having them implies they are supported and have valid values, which they are not. Besides framework rarely expose them as in Obj-C it's much better to check for the existence of APIs vs the version of libraries.

swisspol commented on GCDWebServer/Framework/GCDWebServers.h:
```diff
@@ -0,0 +1,56 @@
+FOUNDATION_EXPORT double GCDWebServersVersionNumber;
+
+//! Project version string for GCDWebServers.
+FOUNDATION_EXPORT const unsigned char GCDWebServersVersionString[];
```
Where is this symbol coming from?

sagmor replied:
Same as above

swisspol commented:
About `Archive.sh`, I'm not sure this is really needed: it's only 2 lines and it's only for the few users of Carthage.

swisspol commented:
About `Archive.sh`, I'm not sure this is really needed: it's only 2 lines and is only useful to the few Carthage (who also presumably know how to use it and don't need the script anyway).

sagmor commented:
They are not really needed by carthage users. I add the Archive script to generate an `official release` build.

Basically whenever you tag a release on GitHub, you run `./Archive.sh` and upload the generated GCDWebServers.framework.zip to the release.

This allows anybody to use the framework without the need of building it.

If you think that's not really needed I'll delete it :smile:

swisspol commented:
> Basically whenever you tag a release on GitHub, you run ./Archive.sh and upload the generated GCDWebServers.framework.zip to the release.

I'm not sure I want to do that: it's better in general to recompile from source as you always get the latest compiler, the same used by your app. Can you make it a separate PR, this way it can be merged later if needed, and independently of this work?

sagmor commented:
Of course.

swisspol commented on .gitignore:
```diff
@@ -3,3 +3,4 @@ xcuserdata
 project.xcworkspace
 
 Tests/Payload
+Carthage/Build
```
Is this still needed?

swisspol commented on GCDWebServer/Framework/Info.plist:
```diff
@@ -0,0 +1,20 @@
+	<string>$(PRODUCT_NAME)</string>
+	<key>CFBundlePackageType</key>
+	<string>FMWK</string>
+	<key>CFBundleVersion</key>
```
Didn't we say `CFBundleVersion` should be removed? It's not going to be maintained anyway i.e. I never change the version in Xcode project.

swisspol commented:
One more thing: in the new targets you added to the project, when you go to their "Build Settings", you should have no build settings except the ones specific to the target itself like framework settings. All the settings come from the project settings. See "GCDWebServer (Mac)" for an example:

![image](https://cloud.githubusercontent.com/assets/627915/7458698/58a97d2a-f24b-11e4-8d80-31dc55437144.png)

swisspol commented:
BTW I know you already put quite some work into this PR which is much appreciated, and it is useful so I will merge it. Just wanna make sure these remaining points are addressed first. Thanks again!

sagmor commented:
@swisspol Thanks for that.

I haven't addressed those issues yet because I'm on vacations right now, but will take care of it as soon as I get back :wink:

swisspol commented:
Please also rebase and clean up your PR (concatenate commits as needed to avoid "fix" commits, clean up messages, etc...). The easiest way to do that is likely to use my Git app http://gitup.co :)

guidomb commented:
I was trying to use the fork in Carthage but it is not working. The main problem is that the schema are not shared and you need that in order for Carthage to be able to build the project. On the other hand I checked the project and I cannot find the framework target. There are app targets but no framework targets. Also when I run `carthage build --no-skip-current` the project does not get build. Any ideas? //cc @sagmor

guidomb commented:
Ooops sorry my bad didn't realise that everything was in the `framework` branch. Sorry for not checkig that out. Just added it to my iOS project and it works!

sagmor commented:
@guidomb Awesome, great to hear :)

sagmor commented:
@swisspol There it is, 

I removed those settings and squashed the fix commits.
Let me know if there is anything else.

swisspol commented:
Thanks but why did you keep 0a87b39db01983355ca211a158fa39ebc1f832c1? Just remove both this one and the revert one :)

Finally, you still have my comment above on 364ac9d0939bfee60c0b8ae86b2eace2ba49ce9f that is waiting on your feedback.

sagmor commented:
Sorry I didn't saw that commit :stuck_out_tongue:

aventurella commented:
@guidomb where is the `framework` branch you speak of?

swisspol commented:
I pulled the frameworks branch in the repo and cleaned up things quite a bit:
https://github.com/swisspol/GCDWebServer/tree/frameworks

If things are OK by people who use Carthage, then I'll merge it in.

guidomb commented:
@swisspol You have to make the last 2 schemes (the one that have the framework targets) shared in order for Carthage to be able to build them.

swisspol commented:
Done. Are they supposed to have some specific configuration settings? Like `Release` instead of the default `Debug` for build? Or does Carthage use the Archive build mode?

guidomb commented:
I am not 100% sure but from make experience making my own libraries Carthage uses the `Realse` mode when building.  For example in one lib I have to add the RELEASE symbol in Release mode in order to make the code compile because I was using an `#ifdef RELEASE` macro statement. 

Is that what you were asking?

guidomb commented:
Just tested in my projects and everything is OK. I think it can be merged into master.

swisspol commented:
OK it's merged. Can someone give me a blurb to add to the READ about how to use GCDWebServer with Carthage? Thx.

guidomb commented:
### [Carthage]
The following would only work after you make a release in github


[Carthage]: https://github.com/Carthage/Carthage

Add the following to your Cartfile:

```
github "swisspol/GCDWebServer"
```

Then run `carthage update`.

Follow the current instructions in [Carthage's README][carthage-installation] for up to date installation instructions.

[carthage-installation]: https://github.com/Carthage/Carthage#adding-frameworks-to-an-application

guidomb commented:
That would is valid after you make a release in github if not people would have to add

```
github "swisspol/GCDWebServer" "master"
```

guidomb commented:
If the next release is going to be `3.2.5` may be you should enforce that in the README by telling people that the minimum version required for Carthage compatibility is `3.2.5`.

```
github "swisspol/GCDWebServer" ~> 3.2.5
```

swisspol commented:
So you've tried from a clean state to just add `github "swisspol/GCDWebServer" "master"` to your cartfile and it worked, right?

guidomb commented:
When I tried it wasn't merged to master yet so I put 

```
github "swisspol/GCDWebServer" "frameworks"
```

swisspol commented:
OK I updated the README for the `master` case. I'll wait until this has been more tested before making it a release. I'm leaving this PR open as a reminder to do a final update of the README.

guidomb commented:
Just tested with `github "swisspol/GCDWebServer" "master"` and it works fine.

swisspol commented:
Version 3.2.5 has been released and README updated accordingly.
